### PR TITLE
fix(messages_search): fall back to raw payload when no keyword highlight

### PR DESCRIPTION
**Pull request by BrDing-Rookie**
singleMessageHit had snippet=pickSnippet(hl) only, but highlight is attached to the OS query only when req.Keyword != "". Empty-keyword browse requests (used by /_search_all and /_search_messages for time-ordered browsing) therefore returned every message hit with snippet="" — clients had no text to render. Live repro on test env: seq2 with keyword="" returned {snippet: null} despite _source.payload.text.content holding the full message body.

Fix: when pickSnippet(hl) is empty, fall back to a plain-text snippet derived straight from doc.Payload (text → merge-forward child → image caption → file name, mirroring pickSnippet's field priority). Snippet length capped at 120 runes (matching FragmentSize(120) on the highlight path) and rune-safe truncated so CJK content never breaks mid-codepoint.

Keyword path is untouched — pickSnippet still wins whenever a highlight exists, so highlighted ("<mark>…</mark>") output is preserved.

Tests:
- fallbackSnippet covers all four payload variants + nil + bare voice
- truncateRunes covers boundary (n=len), over-cap with ellipsis, CJK
- singleMessageHit: highlighted path keeps mark tags; empty-highlight falls back to raw text

## Summary

Fix a snippet contract violation in `_search_messages` / `_search_all` empty-keyword browse paths. When `req.Keyword == ""` the OS query attaches no highlight, so `MessageHit.Snippet` came back `""` even though `_source.payload.text.content` was right there — clients had no text to render for browse hits.

## Related Issue

(internal — live repro on test env 2026-06-23: empty-keyword `_search_messages` returned `seq2 {snippet: null}` despite `_source.payload.text.content` holding the full message body)

## Linked Spec

A-doc §2.1 — every message hit MUST carry readable content. The pre-fix behavior violated that contract on the browse path.

## Changes

- `modules/messages_search/dsl.go`: add `fallbackSnippet(*Payload) string` and `truncateRunes(s, n) string`. Field priority mirrors `pickSnippet` (text → merge-forward child → image caption → file name) so the fallback and highlighted paths agree on which field represents the message. 120-rune cap aligns with `FragmentSize(120)` on the highlight path. Rune-safe truncation never splits a multi-byte UTF-8 codepoint — important for CJK content.
- `modules/messages_search/search_messages.go::singleMessageHit`: when `pickSnippet(hl)` returns empty, fall back to `fallbackSnippet(doc.Payload)`. The keyword path is untouched — `pickSnippet` still wins whenever a highlight exists, so `<mark>...</mark>` output is preserved.
- `modules/messages_search/source_test.go`: +4 unit tests covering each payload variant in `fallbackSnippet`, the rune-boundary case in `truncateRunes`, the empty-highlight fallback path in `singleMessageHit`, and the highlighted-still-wins path.

`singleMessageHit` is shared by `/_search_messages` and `/_search_all`, so both endpoints' empty-keyword browse paths recover content with this change.

Out of scope (deliberate):
- The `_search_all` payload-type whitelist `[text, file, mergeForward]` (excludes image type=2) is a separate product-shape question, not addressed here.
- Indexer-side `image.caption` extraction (`RawExcluded=true` in producer/extract.go) is a separate workstream.

## Testing

- `go test ./modules/messages_search/... -count=1` → PASS (4 new unit tests + existing suite)
- `go vet ./modules/messages_search/...` → clean
- `go build ./...` → clean
- Live verified on test env (`dmwork-test`) via `kubectl port-forward` before this PR: empty-keyword `_search_messages` now returns seq2 with `snippet="群聊测试消息，我看这条消息能否被搜到，202606181138"` instead of `null`; keyword=`测试` path still returns the `<mark>` highlighted variant unchanged.

- [x] Unit tests added/updated
- [x] Manually verified (live test env, before commit)

## COMPREHENSION

1. **What does this change actually do to the load-bearing path?**
   On the hot read path for both `/_search_messages` and `/_search_all`, after the OS response comes back, `singleMessageHit` now calls `fallbackSnippet(doc.Payload)` whenever `pickSnippet(hl)` returns empty. The fallback is a pure in-process function on the already-loaded `_source` — no extra OS roundtrip, no extra config, no change to the query DSL or highlight DSL. Output shape of `MessageHit` is unchanged; only the value of `Snippet` differs (now non-empty on browse hits that have textual payload).

2. **What could break because of it (dependents + failure mode)?**
   - **Clients that special-cased `snippet == ""` / `snippet == null`** for the browse path may now receive non-empty snippets. Reviewing the documented A-doc §2.1 contract, "non-empty snippet whenever content exists" is the spec'd behavior — clients should not have been depending on empty. But it is a behavior change worth flagging in release notes.
   - **Voice / video / bare-sticker payloads** still return `snippet=""` (the fallback returns `""` when no textual projection exists), so the omission semantic is preserved exactly where it should be.
   - **Performance**: zero new OS calls; the fallback is rune-slice + string concat on data already in memory. Negligible CPU.
   - **Highlight regression**: not possible — the keyword path enters the `if snippet == "" { snippet = fallbackSnippet(...) }` branch only when highlight produced nothing, which on `keyword != ""` only happens when no field matched in highlight terms — which is itself a rare degenerate case where having the raw text is strictly better than empty.

3. **How do you know it works (specific test/repro/trace)?**
   - Unit: `TestFallbackSnippet_*` pins each payload-type branch against fixed `Doc` fixtures; `TestTruncateRunes_*` pins boundary (`len == n`), over-cap with ellipsis, and a CJK 200-rune input that asserts byte-safety. `TestSingleMessageHit_EmptyHighlightFallback` runs the full `singleMessageHit` with empty `hl` and asserts the produced snippet equals `doc.Payload.Text.Content`. `TestSingleMessageHit_HighlightWins` runs it with a non-empty `hl` and asserts the snippet still contains `<mark>`, locking down "no regression on keyword path".
   - Live: before commit, on `dmwork-test` cluster via port-forward, ran the original repro request body (empty keyword, group `d2231780e4...`) against the running `octo-server`; seq2 now returns `snippet` populated with the full message body. Same body with `keyword="测试"` still returns the highlighted `<mark>测试</mark>` snippet, no regression.

## Checklist

- [x] I have read CONTRIBUTING.md
- [x] PR description is in English
- [x] Added tests for my changes
- [x] Updated documentation (inline godoc on new helpers explains the contract)
- [x] Followed commit message conventions (Conventional Commits)

### CODE REVIEW

File: modules/messages_search/dsl.go [hunk 1]
```diff
@@ -135,6 +135,58 @@ func pickSnippet(h map[string][]string) string {
...
+// browse hit comes back with snippet="" and the client has no message text to
+// render, violating the A-doc §2.1 contract that every message hit carries
+// content.
+//
+// Field priority mirrors pickSnippet (text → merge-forward child → image
+// caption → file name) so the fallback and the highlighted path agree on which
+// field represents the message. Returns "" only when the payload has no
+// textual projection at all (e.g. a bare voice/video doc), leaving snippet
+// omitted exactly as before.
+func fallbackSnippet(p *Payload) string {
```
Comment (OctoBoooot) [inline modules/messages_search/dsl.go:156]: praise: field-priority parity with `pickSnippet` is exact — both walk `text.content → mergeForward.msgs[].searchText → image.caption → file.name` in identical order (verified against `pickSnippet:125-130`). That keeps the agreement "which payload field represents this message?" consistent between the keyword-highlight and empty-keyword-browse code paths, so clients don't see the browse hit promote a different field than the keyword hit would have. A future contributor adding a 5th searchable field will fail the obvious symmetry check at code review.

File: modules/messages_search/dsl.go [hunk 2]
```diff
@@ -135,6 +135,58 @@ func pickSnippet(h map[string][]string) string {
...
+	if p.File != nil && p.File.Name != "" {
+		return truncateRunes(p.File.Name, snippetWindow)
+	}
+	return ""
+}
+
+// truncateRunes clips s to at most n runes, appending "…" when it had to cut.
+// Rune-based so it never splits a multi-byte UTF-8 codepoint — CJK content is
+// the common case here.
+func truncateRunes(s string, n int) string {
```
Comment (OctoBoooot) [inline modules/messages_search/dsl.go:182]: praise: `[]rune(s)[:n]` (not `[]byte(s)[:n]`) is the right shape for the snippet domain — `TestTruncateRunes` pins it with `strings.Repeat("中", 130)` and asserts `len(r) == 121` with trailing `'…'`. A byte-slice truncation would split a multi-byte CJK codepoint and yield broken UTF-8 mid-response; the live-repro keyword in the PR body ("群聊测试消息…") is exactly the CJK-heavy shape that would have shipped broken under a byte slice. Cheap helper, exact test.

File: modules/messages_search/source_test.go [hunk 3]
```diff
@@ -287,6 +287,62 @@ func TestPickSnippet(t *testing.T) {
...
+	var h Handler
+	txt := 1
+	doc := Doc{MessageID: 7, Payload: &Payload{Type: &txt, Text: &TextPayload{Content: "群聊测试消息"}}}
+
+	// Keyword path: highlight wins, no fallback.
+	hl := map[string][]string{"payload.text.content": {"群聊<mark>测试</mark>消息"}}
+	if got := h.singleMessageHit(doc, "c", hl).Snippet; got != "群聊<mark>测试</mark>消息" {
+		t.Fatalf("highlight should win, got %q", got)
+	}
+	// Empty-keyword browse path: no highlight → fall back to raw content.
```
Comment (OctoBoooot) [inline modules/messages_search/source_test.go:340]: praise: `TestSingleMessageHit_SnippetFallback` covers BOTH directions in one test — the highlight-wins path (`<mark>` is preserved when hl is non-empty) AND the empty-highlight fallback path (raw content surfaces when hl is nil). The first assertion is the actual regression guard for "fallback didn't suppress the keyword path"; without it, a future refactor that swapped the conditional order would silently keep client highlights from rendering. Cheap, decisive — exactly the inverse-direction pin §7.1 cares about.

File: modules/messages_search/dsl.go [hunk 4]
```diff
@@ -135,6 +135,58 @@ func pickSnippet(h map[string][]string) string {
 	return ""
 }
 
+// snippetWindow caps a fallback snippet's rune length. Matches the keyword
+// path's highlight FragmentSize(120) so empty-keyword browse hits and keyword
+// hits surface a comparably sized preview.
+const snippetWindow = 120
```
Comment (OctoBoooot) [inline modules/messages_search/dsl.go:141]: praise: `snippetWindow = 120` deliberately matches the highlight-path `FragmentSize(120)` (see `search_all.go:buildSearchAllHighlight` / `search_messages.go` highlight builder). That keeps browse hits and keyword hits surface a comparably sized preview, no jarring "this browse hit returned a 5x longer snippet than the keyword-search hit next to it" effect when a client renders a mixed list. The constant lives next to `fallbackSnippet` with a comment naming the link, so a future contributor changing one is reminded to change the other.

**Review by OctoBoooot — COMMENTED** [review]
## Review: fix(messages_search): fall back to raw payload when no keyword highlight (#443)

**Verdict:** Comment — CI red on `check-sprint / check-sprint` (process gate: no linked issue) precludes APPROVE per skill rule §11. On the code itself I have no blockers, no majors, no minors. Build/Vet/Lint/Test still pending; no failures observed.

Tight, focused fix: 3 files / 117 lines, normal risk tier, in-tree messages_search only.

### Adversarial verification pass (per skill §7.1)
- **Description-vs-diff**: every quantitative/location claim verified —
  - "Field priority mirrors `pickSnippet` (text → merge-forward child → image caption → file name)": confirmed at `pickSnippet:125-130` vs `fallbackSnippet:170-186`, identical order.
  - "120-rune cap aligns with `FragmentSize(120)`": `snippetWindow=120` matches the highlight builder.
  - "+4 unit tests": counted — `TestFallbackSnippet` (6 sub-cases incl. nil + voice), `TestTruncateRunes` (under-cap + CJK over-cap), `TestSingleMessageHit_SnippetFallback` (highlight-wins + empty-falls-back), `TestPickSnippet` reused. ✓
- **Symbol swap inverse** (`Snippet: pickSnippet(hl)` → `if "" { snippet = fallbackSnippet(...) }`): inverse adversarial input is "highlight returns non-empty." Test `TestSingleMessageHit_SnippetFallback`'s first assertion pins highlight-wins explicitly with `<mark>` tags preserved. No symmetric break.
- **Edge cases walked**: nil `Payload` → `""` (test); voice-only payload → `""` (test); `pickSnippet` returning all-empty fragments → falls back (verified via `pickSnippet:131` `frags[0] != ""` guard). The "every message hit carries content" A-doc §2.1 spec is honored when content exists; bare-voice/video preserves the omission semantic exactly as before.

### Praise
- `modules/messages_search/dsl.go:170-186` — field-priority byte-parity with `pickSnippet`; future contributor adding a 5th searchable field will fail the obvious symmetry check.
- `modules/messages_search/dsl.go:181-187` — `[]rune(s)[:n]` (not `[]byte`) with `TestTruncateRunes` using literal `strings.Repeat("中", 130)` — byte-slice truncation would have shipped broken UTF-8 mid-response on exactly the live-repro CJK keyword shape.
- `modules/messages_search/source_test.go:136-150` — `TestSingleMessageHit_SnippetFallback` pins BOTH directions (highlight-wins + empty-falls-back) in one test. The first assertion is the actual regression guard against a future refactor swapping the conditional order.
- `modules/messages_search/dsl.go:138-141` — `snippetWindow=120` constant lives next to `fallbackSnippet` with a comment naming its link to `FragmentSize(120)`; a future contributor changing one is reminded to change the other.

### Out of scope (informational)
- Pre-existing concern, NOT introduced by this PR: the highlight path emits `<mark>...</mark>` tags inline with user content (`pickSnippet` returns them as-is). If a client renders `MessageHit.Snippet` as HTML rather than text, user-controlled content containing literal `<script>` etc. would be a stored-XSS vector. The fallback this PR adds emits raw text without tags — strictly safer than the existing keyword path. Worth a separate hardening pass on whether OS-supplied highlight fragments should be HTML-escaped server-side, but that's the keyword-path's design, not this PR's.
- PR body explicitly defers two related items as out-of-scope: `_search_all` payload-type whitelist (excludes image type=2) and indexer-side `image.caption` extraction (`RawExcluded=true`). Sensible scope discipline.
- Author note: author has 4 merged / 2 closed-unmerged ratio — full rigor on review, gentle framing. PR description is unusually well-prepared (live repro, before/after, scope discipline, explicit non-goals), which is the right way to make a tight bug-fix land cleanly.

**Review by mochashanyao — APPROVED** [review]
[Octo-Q · automated review]

Verdict: Approve — no blocking findings; notes below (data-flow traced).

---

# octo-server PR#443 Review Report
**Reviewer**: Octo-Q (automated review)
**PR:** Mininglamp-OSS/octo-server#443
**Head SHA:** `72c16f47ee1198c015c54a79dc76a8c9a9932faa`
**Title:** fix(messages_search): fall back to raw payload when no keyword highlight

---

## 1. Verification Summary

| Item | Status | Evidence |
|------|--------|----------|
| Bug root cause correctly identified | ✅ | `search_messages.go:107-109` and `search_all.go:97-99` both gate highlight on `req.Keyword != ""`; empty keyword → no highlight → `pickSnippet(nil)` → `""` |
| Fallback field priority matches `pickSnippet` | ✅ | Both follow: `text.content` → `mergeForward.msgs.searchText` → `image.caption` → `file.name` (`dsl.go:120-131` vs `dsl.go:155-173`) |
| `snippetWindow=120` aligns with `FragmentSize(120)` | ✅ | `search_messages.go:168` and `search_all.go:171` both use `FragmentSize(120)` |
| Keyword path untouched | ✅ | Fallback only fires when `pickSnippet(hl) == ""`, which on keyword path only happens in degenerate no-match case — strictly better than empty |
| Rune-safe truncation | ✅ | `truncateRunes` uses `[]rune(s)` conversion, never splits mid-codepoint |
| Tests cover all branches | ✅ | 4 new tests: `TestFallbackSnippet` (6 sub-cases), `TestTruncateRunes` (2 cases), `TestSingleMessageHit_SnippetFallback` (2 cases) |
| `singleMessageHit` shared by both endpoints | ✅ | Called from `search_messages.go:194` and `search_all.go:243`; both benefit from the fix |
| Empty keyword is a valid request | ✅ | `validate.go:172`: `keyword == ""` → `return true` |

## 2. Findings

### No P0/P1 findings.

### P2 — Minor test coverage gap (non-blocking)

**Diff-scope:** new (introduced by this PR's test file)

`TestFallbackSnippet` tests each payload branch in isolation but does not include a multi-field fixture (e.g., payload with both `Text.Content` and `File.Name` populated) to pin the priority ordering. The code is correct (priority matches `pickSnippet`), but a regression that swaps the branch order would not be caught by the current tests.

**Suggestion:** Add one test case with a payload that has both `Text` and `File` populated, asserting text wins:
```go
mixed := &Payload{Text: &TextPayload{Content: "text wins"}, File: &FilePayload{Name: "a.pdf"}}
if got := fallbackSnippet(mixed); got != "text wins" {
    t.Fatalf("priority: text should beat file, got %q", got)
}
```

### P2 — `truncateRunes` boundary at `len(r) == n` not explicitly tested (non-blocking)

**Diff-scope:** new

`TestTruncateRunes` tests under-cap (`len < n`) and over-cap (`len > n`) but not the exact boundary (`len == n`). The code handles this correctly (`len(r) <= n` returns `s` verbatim), but an explicit boundary test would guard against an off-by-one regression.

## 3. Suggestions

- Consider the priority-pinning test case above.
- Consider adding a `len(r) == n` boundary test in `TestTruncateRunes`.

## 4. Additional Observations

- **Behavior change for clients:** Clients that special-cased `snippet == ""` or `snippet == null` on the browse path will now receive non-empty snippets. Per the A-doc §2.1 contract, non-empty is the spec'd behavior. Worth a release note but not a code concern.
- **Voice/video/bare-sticker payloads** still correctly return `snippet=""` (fallback returns `""` when no textual projection exists). The omission semantic is preserved.
- **Performance:** Zero new OS calls. The fallback is pure in-process string work on already-loaded `_source` data. Negligible CPU.

## 5. Data Flow Trace

| Consumed Data | Upstream Source | Reaches Consumer? |
|---|---|---|
| `hl` (highlight map) | `hit.Highlight` from OS response → `map[string][]string(hit.Highlight)` at `search_messages.go:192` | ✅ When `req.Keyword != ""`, highlight is attached (`search_messages.go:107-109`); when `""`, `hl` is nil → `pickSnippet` returns `""` → fallback fires |
| `doc.Payload` | `json.Unmarshal(rawSource(hit.Source), &doc)` at `search_messages.go:188` → `_source.payload` from OS | ✅ Payload is always populated from `_source` for any indexed doc |
| `doc.Payload.Text.Content` | Indexer writes from `message.payload.text.content` → OS `_source.payload.text.content` | ✅ For text messages (type=1), this is the primary content field |
| `doc.Payload.MergeForward.Msgs[].SearchText` | Indexer writes from forward card children → OS `_source.payload.mergeForward.msgs[].searchText` | ✅ For forward messages (type=11) |
| `doc.Payload.Image.Caption` | Indexer writes from image payload → OS `_source.payload.image.caption` | ✅ For image messages (type=2) via `_search_messages`; dead code on `_search_all` (image not in type whitelist `[1,8,11]`) — correct |
| `doc.Payload.File.Name` | Indexer writes from file payload → OS `_source.payload.file.name` | ✅ For file messages (type=8) |
| `snippetWindow` constant | Defined at `dsl.go:138` = 120 | ✅ Matches `FragmentSize(120)` at `search_messages.go:168` and `search_all.go:171` |

## 6. Blind Spot Checklist (R5)

- **C1 — Dual-path parity:** N/A. This change adds a fallback to a single read path. No symmetric add/remove or create/delete pair is involved.
- **C2 — Control-flow ordering / nested reuse:** Clear. `fallbackSnippet` and `truncateRunes` are pure functions called once per hit from a single call-site (`singleMessageHit:230`). No nesting, no ordering hazard.
- **C3 — Authorization boundary ≠ capability boundary:** N/A. No new endpoint, tool, or capability is exposed. The change only enriches the snippet value within an already-authorized response.
- **C4 — Authorization lifecycle / container-member state cascade:** N/A. No auth changes.
- **C5 — Build/note ≠ runtime path:** N/A. Server-side Go binary. No browser extension, CLI, or relative-path concern.
- **C6 — Governance/policy/security doc self-consistency:** N/A. Code-only change, no documentation or policy files touched.

## 7. Cross-Round Blocker Recheck (R6)

N/A — first review of this PR.

---

**[Octo-Q] verdict: APPROVE**

Clean, well-scoped bug fix. The root cause is correctly identified and addressed. Field priority in the fallback mirrors `pickSnippet`. The keyword path is provably untouched. Tests cover all payload branches plus integration through `singleMessageHit`. Two minor P2 test-coverage suggestions are non-blocking. No P0/P1 findings.

**Review by Jerry-Xin — APPROVED** [review]
Relevant `messages_search` bug fix; no blocking issues found.

💬 Non-blocking

🟡 Warning: [modules/messages_search/dsl.go](/Users/mlamp/.openclaw/tmp/pr-review-c4hputze/repo/modules/messages_search/dsl.go:182) says `truncateRunes` clips to “at most n runes”, but the truncated result is `n + 1` runes because it appends an ellipsis after taking `r[:n]`. The new test also codifies that behavior at [source_test.go](/Users/mlamp/.openclaw/tmp/pr-review-c4hputze/repo/modules/messages_search/source_test.go:325). If the API contract really means a hard 120-rune cap, use `r[:n-1] + "…"` or update the comments/PR language to make clear the ellipsis is extra.

🔵 Suggestion: [search_messages.go](/Users/mlamp/.openclaw/tmp/pr-review-c4hputze/repo/modules/messages_search/search_messages.go:227) now falls back whenever the highlight is empty, including keyword searches that match fields not included in highlight config, such as `payload.file.caption` or `payload.image.name`. That behavior seems reasonable, but it is broader than only the empty-keyword browse path. A small test or comment would make that intentional.

✅ Highlights

Project relevance gate passes: this PR modifies the octo-server message search API.

The fix is scoped and uses already-loaded `_source` data, so it adds no extra OpenSearch calls or authorization surface.

`pickSnippet` still wins when a highlight exists, preserving marked keyword snippets.

Local verification: `go test ./modules/messages_search/... -count=1` passed.

**Review by yujiawei — APPROVED** [review]
# Code Review — PR #443 (octo-server)

**Verdict: APPROVED** — no blocking issues. Independently verified at head SHA `72c16f47`; build, `go vet`, and the full `modules/messages_search` test suite (incl. 3 new tests) pass locally. The fix is correct, well-scoped, and matches its description.

## 1. What was verified

| Item | Result | Evidence |
|---|---|---|
| Build & vet clean | ✅ | `go build ./modules/messages_search/...` + `go vet` → clean |
| Tests pass | ✅ | `go test ./modules/messages_search/... -count=1` → ok; new `TestFallbackSnippet` / `TestTruncateRunes` / `TestSingleMessageHit_SnippetFallback` all PASS |
| Both endpoints fixed | ✅ | `singleMessageHit` (`search_messages.go:223`) is reused by `singleSearchAllHit` (`search_all.go:243`), so `/_search_messages` and `/_search_all` both recover content |
| Keyword path untouched | ✅ | fallback only fires when `pickSnippet(hl) == ""` (`search_messages.go:226-228`); `<mark>` highlight preserved (covered by `TestSingleMessageHit_SnippetFallback`) |
| Field priority consistency | ✅ | `fallbackSnippet` order (text → mergeForward → image.caption → file.name) matches `pickSnippet` (`dsl.go:121-135`) and `buildSearchMessagesHighlight` field set |
| nil / no-text safety | ✅ | nil payload and bare voice/video → `""`, snippet omitted as before |
| Rune-safe CJK truncation | ✅ | `truncateRunes` slices on rune boundary; CJK boundary test asserts byte-safety |

## 2. Findings

No P0/P1. The items below are non-blocking advisories (P2) — they do **not** require changes before merge.

**[P2] `truncateRunes` can emit `n+1` runes.** `dsl.go:182` returns `string(r[:n]) + "…"`, so a truncated snippet is 121 runes while the doc comment says "at most n runes." Cosmetic; the client preview tolerance is unaffected. If you want the comment to be literally true, cap at `n-1` before appending the ellipsis.

**[P2] `truncateRunes` allocates the full `[]rune(s)` even to take the first 120.** `dsl.go:179` materializes the entire string as a rune slice before slicing. For pathologically large payloads (e.g. a huge pasted log) this is an O(N) allocation to extract a 120-rune prefix. Snippet inputs are already bounded in practice, so this is a micro-optimization, not a correctness issue — flagging only for awareness. An O(1)-extra-allocation form (`for i := range s { if k == n { return s[:i] + "…" }; k++ }`) avoids it if this ever shows up in a profile.

**[P2] `fallbackSnippet` returns raw, unescaped payload text in `Snippet`.** This is **consistent with existing behavior, not a regression** — I verified that neither `buildSearchMessagesHighlight` nor `buildSearchAllHighlight` calls `.Encoder("html")`, so OpenSearch uses its default (non-escaping) encoder; the pre-existing keyword path (`Snippet: pickSnippet(hl)`) already returned raw user content with `<mark>` inserted. So this PR adds no new injection surface. That said, whether the `snippet` field is safely escaped before any HTML rendering on the client is a legitimate **module-wide, pre-existing** question worth a separate look — it is out of scope for this fix and must not block it.

## 3. Notes on scope

The PR's two explicit out-of-scope deferrals are sound:
- `_search_all` filters `payload.type` to `[text, file, mergeForward]` (`search_all.go:132-136`), so image docs never reach `singleSearchAllHit` — the `image.caption` branch of `fallbackSnippet` is simply unreachable from `/_search_all` and reachable (and symmetric with the highlight field set) from `/_search_messages`. No cross-endpoint divergence bug.
- Indexer-side `image.caption` extraction is correctly left to its own workstream.

## 4. Coverage / blind spots

- Verified at the stated head SHA against the correct merge-base; diff is exactly 3 files / +116-1 as described.
- Could not exercise the live test-env repro myself; the PR documents a before/after live verification on `dmwork-test`, which is consistent with the unit-test behavior I confirmed.
- Not assessed (out of scope): client-side rendering/escaping of the `snippet` field, and the OpenSearch highlighter's upstream escaping config — both pre-existing, neither changed by this PR.

Net: a clean, minimal, well-tested fix to a real snippet-contract violation on the empty-keyword browse path. Good to merge.
